Tarantool errors are lost if driver calls CompletableFuture.get() with timeout args

TarantoolClient.callForSingleResult() returns CompletableFuture

**get() without timeout args**: TarantoolFunctionCallException with error info from Tarantool
Caused by: TarantoolFunctionCallException: {"type":"ClientError","message":"...' ...}
        at io.tarantool.driver.api.SingleValueCallResultImpl.<init>(SingleValueCallResultImpl.java:28)
        at io.tarantool.driver.mappers.SingleValueCallResultConverter.fromValue(SingleValueCallResultConverter.java:23)
        at io.tarantool.driver.mappers.SingleValueCallResultConverter.fromValue(SingleValueCallResultConverter.java:13)
        at io.tarantool.driver.mappers.DefaultMessagePackMapper.fromValue(DefaultMessagePackMapper.java:107)
        at io.tarantool.driver.mappers.DefaultMessagePackMapper.fromValue(DefaultMessagePackMapper.java:87)
        at io.tarantool.driver.mappers.AbstractResultMapper.fromValue(AbstractResultMapper.java:27)
        at io.tarantool.driver.handlers.TarantoolResponseHandler.channelRead0(TarantoolResponseHandler.java:51)
        at io.tarantool.driver.handlers.TarantoolResponseHandler.channelRead0(TarantoolResponseHandler.java:23)

**get() with timeout args**: no error info from Tarantool
Caused by: java.util.concurrent.TimeoutException: null
	at java.base/java.util.concurrent.CompletableFuture.timedGet(CompletableFuture.java:1886) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2021) ~[na:na]

--- a/src/main/java/io/tarantool/driver/exceptions/TarantoolAttemptsLimitException.java
+++ b/src/main/java/io/tarantool/driver/exceptions/TarantoolAttemptsLimitException.java
@@ -1,0 +1,28 @@
+package io.tarantool.driver.exceptions;
+
+import io.tarantool.driver.retry.TarantoolRequestRetryPolicies;
+
+/**
+ * The exception that was thrown from {@link TarantoolRequestRetryPolicies}
+ *
+ * @author Artyom Dubinin
+ */
+public class TarantoolAttemptsLimitException extends TarantoolException {
+    private static final String message = "Attempts limit reached: %s";
+
+    public TarantoolAttemptsLimitException(Throwable cause) {
+        super(cause);
+    }
+
+    public TarantoolAttemptsLimitException(Integer limit) {
+        super(String.format(message, limit));
+    }
+
+    public TarantoolAttemptsLimitException(Integer limit, Throwable cause) {
+        super(String.format(message, limit), cause);
+    }
+
+    public TarantoolAttemptsLimitException(String format, Object... args) {
+        super(String.format(format, args));
+    }
+}

--- a/src/main/java/io/tarantool/driver/exceptions/TarantoolTimeoutException.java
+++ b/src/main/java/io/tarantool/driver/exceptions/TarantoolTimeoutException.java
@@ -1,0 +1,28 @@
+package io.tarantool.driver.exceptions;
+
+import io.tarantool.driver.retry.TarantoolRequestRetryPolicies;
+
+/**
+ * The exception that was thrown from {@link TarantoolRequestRetryPolicies}
+ *
+ * @author Artyom Dubinin
+ */
+public class TarantoolTimeoutException extends TarantoolException {
+    private static final String message = "Operation timeout value exceeded after %s ms";
+
+    public TarantoolTimeoutException(Throwable cause) {
+        super(cause);
+    }
+
+    public TarantoolTimeoutException(Long time) {
+        super(String.format(message, time));
+    }
+
+    public TarantoolTimeoutException(Long time, Throwable cause) {
+        super(String.format(message, time), cause);
+    }
+
+    public TarantoolTimeoutException(String format, Object... args) {
+        super(String.format(format, args));
+    }
+}

--- a/src/main/java/io/tarantool/driver/retry/RequestRetryPolicy.java
+++ b/src/main/java/io/tarantool/driver/retry/RequestRetryPolicy.java
@@ -1,6 +1,7 @@
 package io.tarantool.driver.retry;
 
 import io.tarantool.driver.ConnectionSelectionStrategy;
+import io.tarantool.driver.exceptions.TarantoolTimeoutException;
 import io.tarantool.driver.utils.Assert;
 
 import java.util.concurrent.CompletableFuture;

--- a/src/main/java/io/tarantool/driver/retry/TarantoolRequestRetryPolicies.java
+++ b/src/main/java/io/tarantool/driver/retry/TarantoolRequestRetryPolicies.java
@@ -1,6 +1,8 @@
 package io.tarantool.driver.retry;
 
+import io.tarantool.driver.exceptions.TarantoolAttemptsLimitException;
 import io.tarantool.driver.exceptions.TarantoolClientException;
+import io.tarantool.driver.exceptions.TarantoolTimeoutException;
 import io.tarantool.driver.utils.Assert;
 
 import java.util.concurrent.CompletableFuture;
@@ -16,6 +18,7 @@ import java.util.function.Supplier;
  * Class-container for built-in request retry policies
  *
  * @author Alexey Kuzin
+ * @author Artyom Dubinin
  */
 public final class TarantoolRequestRetryPolicies {
 
@@ -44,7 +47,7 @@ public final class TarantoolRequestRetryPolicies {
          * @param requestTimeout   timeout for one retry attempt, in milliseconds
          * @param operationTimeout timeout for the whole operation, in milliseconds
          * @param delay            delay between attempts, in milliseconds
-         * @param exceptionCheck         function checking whether the given exception may be retried
+         * @param exceptionCheck   function checking whether the given exception may be retried
          */
         public InfiniteRetryPolicy(long requestTimeout, long operationTimeout, long delay, T exceptionCheck) {
             Assert.state(requestTimeout >= 0, "Timeout must be greater or equal than 0!");
@@ -104,7 +107,9 @@ public final class TarantoolRequestRetryPolicies {
                     }
                     timeElapsed = timeElapsed + (System.nanoTime() - tStart) / 1_000_000L;
                     if (timeElapsed >= getOperationTimeout()) {
-                        ex = new TimeoutException("Operation timeout value exceeded after " + timeElapsed + " ms");
+                        ex = new TarantoolTimeoutException(
+                                timeElapsed,
+                                ex);
                         break;
                     }
                 } while (this.canRetryRequest(ex));
@@ -226,6 +231,7 @@ public final class TarantoolRequestRetryPolicies {
             implements RequestRetryPolicy {
 
         private int attempts;
+        private final int limit;
         private final long requestTimeout; //ms
         private final long delay; //ms
         private final T exceptionCheck;
@@ -250,6 +256,7 @@ public final class TarantoolRequestRetryPolicies {
             Assert.notNull(exceptionCheck, "Exception checking callback must not be null!");
 
             this.attempts = attempts;
+            this.limit = attempts;
             this.requestTimeout = requestTimeout;
             this.delay = delay;
             this.exceptionCheck = exceptionCheck;
@@ -269,6 +276,33 @@ public final class TarantoolRequestRetryPolicies {
                 return true;
             }
             return false;
+        }
+
+        @Override
+        public <T> CompletableFuture<T> wrapOperation(Supplier<CompletableFuture<T>> operation, Executor executor) {
+
+            Assert.notNull(operation, "Operation must not be null");
+            Assert.notNull(executor, "Executor must not be null");
+
+            return CompletableFuture.supplyAsync(() -> {
+                Throwable ex;
+                do {
+                    try {
+                        return operation.get().get(getRequestTimeout(), TimeUnit.MILLISECONDS);
+                    } catch (TimeoutException | InterruptedException e) {
+                        ex = e;
+                    } catch (ExecutionException e) {
+                        ex = e.getCause();
+                    }
+                    if (attempts == 0) {
+                        ex = new TarantoolAttemptsLimitException(
+                                limit,
+                                ex);
+                        break;
+                    }
+                } while (this.canRetryRequest(ex));
+                throw new CompletionException(ex);
+            }, executor);
         }
     }
 

--- a/src/test/java/io/tarantool/driver/integration/RetryingTarantoolTupleClientIT.java
+++ b/src/test/java/io/tarantool/driver/integration/RetryingTarantoolTupleClientIT.java
@@ -4,7 +4,10 @@ import io.tarantool.driver.ClusterTarantoolTupleClient;
 import io.tarantool.driver.ProxyTarantoolTupleClient;
 import io.tarantool.driver.TarantoolClientConfig;
 import io.tarantool.driver.auth.SimpleTarantoolCredentials;
+import io.tarantool.driver.exceptions.TarantoolAttemptsLimitException;
 import io.tarantool.driver.exceptions.TarantoolFunctionCallException;
+import io.tarantool.driver.exceptions.TarantoolServerException;
+import io.tarantool.driver.exceptions.TarantoolTimeoutException;
 import io.tarantool.driver.retry.RetryingTarantoolTupleClient;
 import io.tarantool.driver.retry.TarantoolRequestRetryPolicies;
 import org.junit.jupiter.api.BeforeAll;
@@ -26,6 +29,7 @@ import static org.junit.jupiter.api.Assertions.fail;
 
 /**
  * @author Alexey Kuzin
+ * @author Artyom Dubinin
  */
 public class RetryingTarantoolTupleClientIT extends SharedCartridgeContainer {
 
@@ -59,7 +63,7 @@ public class RetryingTarantoolTupleClientIT extends SharedCartridgeContainer {
     }
 
     @Test
-    void testSuccessAfterSeveralRetries() throws Exception {
+    void testAttemptsBoundSuccess() throws Exception {
         try (ProxyTarantoolTupleClient client = setupClient()) {
             // The stored function will fail 3 times
             client.call("setup_retrying_function", Collections.singletonList(3));
@@ -70,19 +74,21 @@ public class RetryingTarantoolTupleClientIT extends SharedCartridgeContainer {
     }
 
     @Test
-    void testFailAfterSeveralRetries() throws Exception {
+    void testAttemptsBoundLimitReachedWithTarantoolException() throws Exception {
         try {
             ProxyTarantoolTupleClient client = setupClient();
             client.call("setup_retrying_function", Collections.singletonList(3));
             retrying(client, 2).callForSingleResult("retrying_function", String.class).get();
             fail("Exception must be thrown after last retry attempt.");
         } catch (Throwable e) {
-            assertTrue(e.getCause() instanceof TarantoolFunctionCallException);
+            assertTrue(e.getCause() instanceof TarantoolAttemptsLimitException);
+            assertTrue(e.getCause().getMessage().contains("Attempts limit reached:"));
+            assertTrue(e.getCause().getCause() instanceof TarantoolFunctionCallException);
         }
     }
 
     @Test
-    void testNumberOfAttemptsPolicyFailAfterTimeout() throws Exception {
+    void testAttemptsBoundLimitReachedWithTimeout() throws Exception {
         ProxyTarantoolTupleClient client = setupClient();
 
         client.call("reset_request_counters");
@@ -97,15 +103,36 @@ public class RetryingTarantoolTupleClientIT extends SharedCartridgeContainer {
             retryingClient.call("long_running_function", Collections.singletonList(10)).get();
             fail("Exception must be thrown after timeout.");
         } catch (ExecutionException e) {
-            assertTrue(e.getCause() instanceof TimeoutException, "Unexpected exception type: " + e);
+            assertTrue(e.getCause() instanceof TarantoolAttemptsLimitException, "Unexpected exception type: " + e);
+            assertTrue(e.getCause().getCause() instanceof TimeoutException, "Unexpected exception type: " + e);
         }
+    }
+
+    @Test
+    void testInfiniteRetrySuccess() throws Exception {
+        ProxyTarantoolTupleClient client = setupClient();
+
+        RetryingTarantoolTupleClient retryingClient = new RetryingTarantoolTupleClient(client,
+                TarantoolRequestRetryPolicies
+                        .unbound(e -> {
+                            return e.getMessage().contains("Unsuccessful attempt");
+                        })
+                        .withRequestTimeout(200) //requestTimeout
+                        .withOperationTimeout(2000)  //operationTimeout
+                        .build());
+
+        client.call("setup_retrying_function", Collections.singletonList(3));
+        CompletableFuture<String> f = retryingClient
+                .callForSingleResult("retrying_function", String.class);
+
+        assertEquals("Success", f.get());
     }
 
     /**
      * This test covers https://github.com/tarantool/cartridge-java/issues/83
      */
     @Test
-    void testUnboundPolicyFailsAfterTimeout() throws Exception {
+    void testInfiniteRetryContinuedAfterGetFailed() throws Exception {
         ProxyTarantoolTupleClient client = setupClient();
 
         client.call("reset_request_counters");
@@ -138,7 +165,59 @@ public class RetryingTarantoolTupleClientIT extends SharedCartridgeContainer {
         //check that all worker threads has stopped (counter remain unchanged)
         counterVal = counter.get();
         Thread.sleep(100);
-        assertEquals(counterVal, counter.get(),
-                "Policy continues executing request after timeout");
+        assertEquals(counterVal, counter.get(), "Policy continues executing request after timeout");
+    }
+
+    /**
+     * This test covers https://github.com/tarantool/cartridge-java/issues/82
+     */
+    @Test
+    void testInfiniteRetryTimeoutReachedWithTarantoolException() throws Exception {
+        ProxyTarantoolTupleClient client = setupClient();
+
+        RetryingTarantoolTupleClient retryingClient = new RetryingTarantoolTupleClient(client,
+                TarantoolRequestRetryPolicies
+                        .unbound()
+                        .withRequestTimeout(20) //requestTimeout
+                        .withOperationTimeout(200)  //operationTimeout
+                        .build());
+
+        CompletableFuture<List<?>> f = retryingClient
+                .call("raising_error", Collections.singletonList(10));
+
+        try {
+            f.get();
+        } catch (ExecutionException e) {
+            assertEquals(ExecutionException.class, e.getClass());
+            assertEquals(TarantoolTimeoutException.class, e.getCause().getClass());
+            assertTrue(e.getCause().getMessage().contains("Operation timeout value exceeded after"));
+            assertEquals(TarantoolServerException.class, e.getCause().getCause().getClass());
+            assertTrue(e.getCause().getCause().getMessage().contains("Test error: raising_error() called"));
+        }
+    }
+
+    @Test
+    void testInfiniteRetryTimeoutReachedWithTimeout() throws Exception {
+        ProxyTarantoolTupleClient client = setupClient();
+
+        RetryingTarantoolTupleClient retryingClient = new RetryingTarantoolTupleClient(client,
+                TarantoolRequestRetryPolicies
+                        .unbound()
+                        .withRequestTimeout(20) //requestTimeout
+                        .withOperationTimeout(200)  //operationTimeout
+                        .build());
+
+        client.call("reset_request_counters");
+        CompletableFuture<List<?>> f = retryingClient
+                .call("long_running_function", Collections.singletonList(10));
+
+        try {
+            f.get();
+        } catch (ExecutionException e) {
+            assertEquals(ExecutionException.class, e.getClass());
+            assertEquals(TarantoolTimeoutException.class, e.getCause().getClass());
+            assertTrue(e.getCause().getMessage().contains("Operation timeout value exceeded after"));
+            assertEquals(TimeoutException.class, e.getCause().getCause().getClass());
+        }
     }
 }

--- a/src/test/java/io/tarantool/driver/retry/RequestRetryPolicyTest.java
+++ b/src/test/java/io/tarantool/driver/retry/RequestRetryPolicyTest.java
@@ -1,5 +1,6 @@
 package io.tarantool.driver.retry;
 
+import io.tarantool.driver.exceptions.TarantoolAttemptsLimitException;
 import io.tarantool.driver.exceptions.TarantoolClientException;
 import org.junit.jupiter.api.Test;
 
@@ -104,8 +105,10 @@ class RequestRetryPolicyTest {
             wrappedFuture.get();
         } catch (ExecutionException e) {
             thrown = e;
-            assertEquals(RuntimeException.class, e.getCause().getClass());
-            assertEquals("Should fail 1 times", e.getCause().getMessage());
+            assertEquals(TarantoolAttemptsLimitException.class, e.getCause().getClass());
+            assertEquals("Attempts limit reached: 3", e.getCause().getMessage());
+            assertEquals(RuntimeException.class, e.getCause().getCause().getClass());
+            assertEquals("Should fail 1 times", e.getCause().getCause().getMessage());
         }
         assertNotNull(thrown, "No exception has been thrown");
     }
@@ -121,8 +124,10 @@ class RequestRetryPolicyTest {
             wrappedFuture.get();
         } catch (ExecutionException e) {
             thrown = e;
-            assertEquals(RuntimeException.class, e.getCause().getClass());
-            assertEquals("Should fail 1 times", e.getCause().getMessage());
+            assertEquals(TarantoolAttemptsLimitException.class, e.getCause().getClass());
+            assertEquals("Attempts limit reached: 0", e.getCause().getMessage());
+            assertEquals(RuntimeException.class, e.getCause().getCause().getClass());
+            assertEquals("Should fail 1 times", e.getCause().getCause().getMessage());
         }
         assertNotNull(thrown, "No exception has been thrown");
     }
@@ -134,8 +139,10 @@ class RequestRetryPolicyTest {
         try {
             wrappedFuture.get();
         } catch (ExecutionException e) {
-            assertTrue(e.getCause() instanceof RuntimeException);
-            assertEquals("Fail", e.getCause().getMessage());
+            assertTrue(e.getCause() instanceof TarantoolAttemptsLimitException);
+            assertEquals("Attempts limit reached: 4", e.getCause().getMessage());
+            assertTrue(e.getCause().getCause() instanceof RuntimeException);
+            assertEquals("Fail", e.getCause().getCause().getMessage());
         }
     }
 

